gpflow/experimental/check_shapes/error_contexts.py

```diff
@@ -0,0 +1,436 @@
+            elif isinstance(line, tuple):
+                fragments.append(self._indent_str)
+                for i, t in enumerate(line):
+                    fragments.append(t)
+                    if i + 1 < len(line):
+                        fragments.append(" " * (self._column_widths[i] - len(t) + 1))
+                fragments.append("\n")
+            else:
+                assert isinstance(line, MessageBuilder)
+                line._collect(fragments)  # pylint: disable=protected-access
+
+    def build(self) -> str:
+        """
+        Compile all collected text into a single string.
+        """
+        fragments: List[str] = []
+        self._collect(fragments)
+        return "".join(fragments)
+
+
+class ErrorContext(ABC):
+    """
+    A context in which an error can occur.
+
+    Contexts should be immutable, and implement __eq__ - so that they can be composed using
+    `StackContext` and `ParallelContext`.
+
+    The contexts are often created even if an error doesn't actually occur, so they should be cheap
+    to create - prefer to do any slow computation in `format`, rather than in `__init__`.
+
+    Maybe think of an `ErrorContext` as a factory of error messages.
+    """
+
+    @abstractmethod
+    def format(self, builder: MessageBuilder) -> None:
+        """
+        Print this context to the given `MessageBuilder`.
+        """
+
+
+@dataclass(frozen=True)
+class StackContext(ErrorContext):
+    """
+    Error context where one context is "inside" another one.
+    """
+
+    parent: ErrorContext
+    child: ErrorContext
+
+    def format(self, builder: MessageBuilder) -> None:
+        flat: List[ErrorContext] = []
+        self._flatten(self, flat)
+        self._format(builder, flat, 0)
+
+    def _flatten(self, context: ErrorContext, flat: List[ErrorContext]) -> None:
+        if isinstance(context, StackContext):
+            self._flatten(context.parent, flat)
+            self._flatten(context.child, flat)
+        else:
+            flat.append(context)
+
+    def _format(self, builder: MessageBuilder, flat: Sequence[ErrorContext], i: int) -> None:
+        if i < len(flat):
+            flat[i].format(builder)
+            with builder.indent():
+                self._format(builder, flat, i + 1)
+
+
+@dataclass(frozen=True)
+class ParallelContext(ErrorContext):
+    """
+    Error context with many contexts in parallel.
+    """
+
+    children: Sequence[ErrorContext]
+
+    def format(self, builder: MessageBuilder) -> None:
+        flat: List[ErrorContext] = []
+        # Merge any nested parallel contexts.
+        self._flatten(self, flat)
+
+        # If any `StackContexts`s have the same (grand-) parent, merge them, so that we don't repeat
+        # stuff in the output.
+        by_head: Dict[ErrorContext, List[ErrorContext]] = {}
+        for child in flat:
+            if isinstance(child, StackContext):
+                head, body = self._split_head(child)
```

Comment:
(can you not just `._flatten` child to get the head, since you'll be doing that later anyway?)